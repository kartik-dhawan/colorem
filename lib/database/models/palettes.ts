// imports 'palettes' collection from mongo db atlas
import mongoose from "mongoose"
const Schema = mongoose.Schema

const PaletteSchema = new Schema(
  {
    result: [[Number]],
  },
  { collection: "palettes" }
)

export default mongoose.models.palettes ||
  mongoose.model("palettes", PaletteSchema)
