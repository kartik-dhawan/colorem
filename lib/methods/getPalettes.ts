import colorPalette from "../database/models/colorPalette"
import { FinalPaletteType } from "../utils/interfaces"

// this function returns a promise which further consists the response.
export const getAllColorPalettes: () => Promise<
  FinalPaletteType[]
> = async () => {
  // .find() finds all the records from the collection
  return await colorPalette.find()
}
