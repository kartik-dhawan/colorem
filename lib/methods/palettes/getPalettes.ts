import axios from "axios"
import { NextApiRequest, NextApiResponse } from "next"
import { logger } from ".."
import colorPalette from "../../database/models/colorPalette"
import { responseTexts, URLS } from "../../utils/constants"
import { FinalPaletteType } from "../../utils/interfaces"

/**
 * this function returns a promise which further consists the response.
//  * @returns {Promise<FinalPaletteType[]>} 
 */
export const getAllColorPalettes: () => Promise<
  FinalPaletteType[]
> = async () => {
  // .find() finds all the records from the collection
  return await colorPalette.find()
}

/**
 *
 * this function fetches all colormind models in a list
 */
export const getColormindModels = async () => {
  return await axios
    .get(URLS.COLORMIND_MODELS_API)
    .then((res) => {
      return res.data
    })
    .catch((err) => {
      logger({
        message: responseTexts.COULD_NOT_COMPLETE.COLORMIND_MODEL,
        error: err,
      })
      return err
    })
}

/**
 * a function to get a palette on the basis of paletteGuid
 * @param {string | string[] | undefined} paletteGuid
 * @param {NextApiRequest} req
 * @param {NextApiResponse} res
 * @returns
 */
export const getPaletteByGuid = async (
  paletteGuid: string | string[] | undefined,
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const filter = {
    paletteGuid: paletteGuid,
  }

  const palette = await colorPalette.findOne(filter)

  if (palette) {
    return palette
  } else {
    res.status(404).json(responseTexts.PALETTE_NOT_FOUND)
  }
}
