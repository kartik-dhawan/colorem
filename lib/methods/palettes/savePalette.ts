import axios from "axios"
import { NextApiRequest, NextApiResponse } from "next"
import colorPalette from "../../database/models/colorPalette"
import { responseTexts, URLS } from "../../utils/constants"
import { getPaletteByGuid } from "./getPalettes"

/**
 * runs colormind api and returns a promise
 * @param {string} model
 * @returns
 */
export const getColormindPalette = async (model: string) => {
  const res = await axios.post(
    URLS.COLORMIND_API,
    { model: model },
    {
      headers: {
        "Content-Type": "text/plain",
      },
    }
  )
  return res.data
}

/**
 * converts an array of colors in RGB to an array of hexcodes
 * @param {number[][]} palette
 * @returns
 */
export const rgbArrayToHex = (palette: number[][]) => {
  const hexArray = palette.map((color) => {
    const hex = color
      .map((code) => {
        return code.toString(16).padStart(2, "0")
      })
      .join("")
    return hex
  })
  return hexArray
}

/**
 * updates palettes likes
 * @param {NextApiRequest} req
 * @param {NextApiResponse} res
 * @param {string | string[] | undefined} paletteGuid
 * @returns
 */
export const updatePaletteByGuid = async (
  req: NextApiRequest,
  res: NextApiResponse,
  paletteGuid: string | string[] | undefined
) => {
  // gets the palette matching the GUID from URL
  const palette = await getPaletteByGuid(paletteGuid, req, res)

  // returns a generic response if no palette with that Guid is found
  if (!palette) {
    res.status(200).json(responseTexts.PALETTE_NOT_FOUND)
  }

  // filter on the basis of which the palette would be searched and then updated
  const filter = {
    paletteGuid: paletteGuid,
  }

  // update object contains the key value pair of the items to be added/updated
  let update
  if (req.body.like === true) {
    // adds a like count
    update = {
      likes: palette?.likes ? palette?.likes + 1 : 1,
    }
  } else if (req.body.like === false) {
    // subtracts a like count
    update = {
      likes: palette?.likes && palette?.likes - 1,
    }
  } else {
    res.status(200).json(responseTexts.BODY_NOT_PRESENT)
  }

  // final updated palette after increasing or decreasing the like count
  const updatedPalette = await colorPalette.findOneAndUpdate(filter, update, {
    new: true, // added so as to return the new updated object after the process
  })

  return updatedPalette
}
