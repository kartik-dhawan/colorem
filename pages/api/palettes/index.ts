// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from "next"
import connectDatabase from "../../../lib/database/connect"
import { getAllColorPalettes } from "../../../lib/methods/getPalettes"

const getPalettes = async (req: NextApiRequest, res: NextApiResponse) => {
  // using POST method even when we just need GET because Next.js doesn't
  // hide our API end points after deployment.
  // ex: colorem.vercel.app/api/palettes would still return the api response
  // when we hit it anywhere, hence exposing our APIs
  if (req.method === "POST") {
    // fetches all the color palettes from the DB
    getAllColorPalettes()
      .then((response) => {
        res.status(200).json(response)
      })
      .catch((err) => {
        console.log({
          message: "Error fetching records.",
          error: err,
        })
        res.status(501).json("Could not fetch color palettes from MongoDB")
      })
  } else {
    res
      .status(403)
      .send("Cannot access this api. Try sending a POST request to this API.")
  }
}

export default connectDatabase(getPalettes)
