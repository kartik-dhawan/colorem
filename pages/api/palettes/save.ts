// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import { v4 as uuidv4 } from "uuid"
import randomWords from "random-words"
import type { NextApiRequest, NextApiResponse } from "next"
import connectDatabase from "../../../lib/database/connect"
import {
  getColormindPalette,
  rgbArrayToHex,
} from "../../../lib/methods/savePalette"
import {
  FinalPaletteType,
  RandomColormindModelType,
} from "../../../lib/utils/interfaces"
import colorPalette from "../../../lib/database/models/colorPalette"
import { getColormindModels } from "../../../lib/methods/getPalettes"

const savePalette = async (req: NextApiRequest, res: NextApiResponse) => {
  // gets a list of available models
  const colormindModels = await getColormindModels()

  // gets a random colormind model-type to fetch a random palette
  const getRandomColormindModel: RandomColormindModelType = () => {
    return colormindModels.result[Math.floor(Math.random() * 6)]
  }

  // using POST method even when we just need GET because Next.js doesn't
  // hide our API end points after deployment.
  // ex: colorem.vercel.app/api/palettes would still return the api response
  // when we hit it anywhere, hence exposing our APIs
  if (req.method === "POST") {
    const randomModel = getRandomColormindModel()
    await getColormindPalette(randomModel)
      .then((response) => {
        // gets array of colors in hex-code
        const paletteInHexCode = rgbArrayToHex(response.result)

        const paletteResponse: FinalPaletteType = {
          paletteGuid: uuidv4(),
          name: randomWords({ exactly: 1, wordsPerString: 2 })[0],
          hex: paletteInHexCode,
          rgb: response.result,
          model: randomModel,
        }

        // creates an instance of the schema
        const palette = new colorPalette(paletteResponse)

        // saves that instance into the DB
        palette
          .save()
          .then(() => {
            // sends the palette as a response to the API if saved
            console.log("============= Saved the palette to DB =============")
            res.status(200).json(paletteResponse)
          })
          .catch(() => {
            // else sends the error response to the API
            res.status(501).json("Could not save the palette to MongoDB")
          })
      })
      .catch((err) => {
        res.status(401).json({
          message: "Could not fetch the palette from ColorMind API",
          err,
        })
      })
  } else {
    res
      .status(403)
      .json("Cannot access this API. Try sending a POST request to this API.")
  }
}

export default connectDatabase(savePalette)
