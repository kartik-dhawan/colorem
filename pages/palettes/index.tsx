import { Box } from "@mui/material"
import PaletteSection from "../../components/PaletteSection"
import { useDispatch } from "react-redux"
import { useEffect } from "react"
import { updatePalettes } from "../../redux/slices/paletteSlice"
import { client } from "../../utils/contentful/config"
import { updateContent } from "../../redux/slices/contentSlice"
import { ContentfulType, PaletteDataType } from "../../utils/interfaces"
import { ErrorBoundary } from "react-error-boundary"
import ErrorFallback, {
  myErrorHandler,
} from "../../components/common/ErrorFallback"
import useSWR from "swr"
import { API_URLS } from "../../utils/constants"
import { logger } from "../../lib/methods"
import { fetcher } from "../../utils/methods"
import { Roboto } from "@next/font/google"

export const getStaticProps = async () => {
  const contentResponse = await client.getEntries({
    content_type: "coloremDashboard", // eslint-disable-line
  })

  return {
    props: {
      contentData: contentResponse?.items[0]?.fields,
    },
    revalidate: parseInt(process.env.ISR_REVAL_TIME_DASHBOARD || "10"), // In seconds
  }
}

const roboto = Roboto({ display: "swap", weight: "300", subsets: ["latin"] })

const Palettes = ({ contentData }: ContentfulType) => {
  const dispatch = useDispatch()

  // using SWR for data fetching because when we use useEffect(), after react 18
  // it calls the function twice.
  // while using useSWR() it only gets called once, and even when we try to
  // call it again in any of the subcomponent it uses the response of its previous call
  const { data, error } = useSWR<PaletteDataType[]>(
    API_URLS.GET_ALL_PALETTES,
    fetcher
  )

  useEffect(() => {
    // if we get data on axios call, store it in redux
    dispatch(updatePalettes(data ? data : []))

    // handle the error here, set it in redux and display a generic error page.
    if (error) {
      // we need to create an error page for error condition
      logger({ error, type: "error" })
    }

    // storing contentful data in redux for this page
    dispatch(updateContent(contentData))
  })

  return (
    <Box
      sx={{
        flex: 1,
        display: "flex",
      }}
      className={roboto.className}
    >
      <ErrorBoundary FallbackComponent={ErrorFallback} onError={myErrorHandler}>
        <PaletteSection />
      </ErrorBoundary>
    </Box>
  )
}

export default Palettes
