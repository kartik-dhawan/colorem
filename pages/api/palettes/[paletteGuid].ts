import type { NextApiRequest, NextApiResponse } from "next"
import { logger } from "../../../lib/methods"
import { getPaletteByGuid } from "../../../lib/methods/getPalettes"
import { updatePaletteByGuid } from "../../../lib/methods/savePalette"

const individualPalette = async (req: NextApiRequest, res: NextApiResponse) => {
  // gets the params from URL
  const { paletteGuid } = req.query

  // PUT - api route-method tp update likes of a palette
  if (req.method === "PUT") {
    // a function to like or unlike the palette
    return req.body.type === "Like/Unlike" // eslint-disable-line
      ? updatePaletteByGuid(req, res, paletteGuid)
          .then((updatedPalette) => {
            res.status(200).json(updatedPalette)
          })
          .catch((err) => {
            logger({ error: err, type: "Error" })
            res.status(501).json({
              message: "Could not complete the PUT request.",
              err,
            })
          })
      : res.status(200).json({
          message: "Please select a type of PUT request in context to the app.",
          types: ["Like/Unlike"],
        })
  }

  // GET - api route-method to get a single palette by Guid
  else if (req.method === "POST") {
    getPaletteByGuid(paletteGuid, req, res)
      .then((palette) => {
        res.status(200).json(palette)
      })
      .catch((err) => {
        logger({ error: err, type: "Error" })
        res.status(501).json({
          message: "Could not complete the PUT request.",
          err,
        })
      })
  } else {
    res
      .status(403)
      .json("Cannot access this API. Try sending a PUT request to this API.")
  }
}

export default individualPalette
