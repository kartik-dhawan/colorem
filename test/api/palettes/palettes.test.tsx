import { getAllColorPalettes } from "../../../lib/methods/palettes/getPalettes"
import { palettesData } from "../../../utils/MockData/palettes"

jest.mock("../../../lib/methods/palettes/getPalettes", () => ({
  getAllColorPalettes: () => {
    return Promise.resolve(palettesData)
  },
}))

describe("Get all palettes data fetch call", () => {
  it("-- should return correct data", async () => {
    try {
      const res = await getAllColorPalettes()
      expect(res).toMatchSnapshot()
    } catch (error) {
      console.log("Get all palettes data fetching is failed", error)
    }
  })
})
