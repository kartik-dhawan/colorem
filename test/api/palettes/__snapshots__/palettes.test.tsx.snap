// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get all palettes data fetch call -- should return correct data 1`] = `
[
  {
    "__v": 0,
    "_id": "639a09c9e9a48bb5f02afa29",
    "hex": [
      "ded798",
      "da9f75",
      "976b50",
      "372f2d",
      "9a986e",
    ],
    "likes": 3,
    "model": "makoto_shinkai",
    "name": "ocean than",
    "paletteGuid": "e0c00261-a904-430c-9777-3232d9173c59",
    "rgb": [
      [
        222,
        215,
        152,
      ],
      [
        218,
        159,
        117,
      ],
      [
        151,
        107,
        80,
      ],
      [
        55,
        47,
        45,
      ],
      [
        154,
        152,
        110,
      ],
    ],
  },
  {
    "__v": 0,
    "_id": "639a0ac9e9a48bb5f02afa2b",
    "hex": [
      "788db4",
      "3f3b68",
      "5f6b81",
      "aaa38f",
      "adb995",
    ],
    "likes": 3,
    "model": "makoto_shinkai",
    "name": "coffee slope",
    "paletteGuid": "c8e9fefd-515c-4c3a-9b32-595b07f252f9",
    "rgb": [
      [
        120,
        141,
        180,
      ],
      [
        63,
        59,
        104,
      ],
      [
        95,
        107,
        129,
      ],
      [
        170,
        163,
        143,
      ],
      [
        173,
        185,
        149,
      ],
    ],
  },
  {
    "__v": 0,
    "_id": "639a0ad3e9a48bb5f02afa2d",
    "hex": [
      "44381f",
      "7d8230",
      "becf8f",
      "69b388",
      "629a8b",
    ],
    "model": "the_wind_rises",
    "name": "jump exercise",
    "paletteGuid": "6d91d4b1-cd78-4503-8ecd-3525f5f8e8bb",
    "rgb": [
      [
        68,
        56,
        31,
      ],
      [
        125,
        130,
        48,
      ],
      [
        190,
        207,
        143,
      ],
      [
        105,
        179,
        136,
      ],
      [
        98,
        154,
        139,
      ],
    ],
  },
  {
    "__v": 0,
    "_id": "639a0c75e9a48bb5f02afa2f",
    "hex": [
      "36251b",
      "845f58",
      "ab9e97",
      "d0cec9",
      "6b6c70",
    ],
    "model": "city_photography",
    "name": "headed oldest",
    "paletteGuid": "549cf550-c198-4452-9ad9-78d80d8f78e3",
    "rgb": [
      [
        54,
        37,
        27,
      ],
      [
        132,
        95,
        88,
      ],
      [
        171,
        158,
        151,
      ],
      [
        208,
        206,
        201,
      ],
      [
        107,
        108,
        112,
      ],
    ],
  },
  {
    "__v": 0,
    "_id": "639a0cace9a48bb5f02afa31",
    "hex": [
      "1d1229",
      "3d5189",
      "2c56ab",
      "6497be",
      "cec0b2",
    ],
    "model": "city_photography",
    "name": "bean aside",
    "paletteGuid": "5022510a-a109-4297-a4ea-4ccb000cb571",
    "rgb": [
      [
        29,
        18,
        41,
      ],
      [
        61,
        81,
        137,
      ],
      [
        44,
        86,
        171,
      ],
      [
        100,
        151,
        190,
      ],
      [
        206,
        192,
        178,
      ],
    ],
  },
]
`;
